report/web.py

```diff
@@ -89,6 +93,115 @@ class Target:
+
+  def listdir(self) -> List[str]:
+    """listdir returns a list of files and directories in path."""
+    if self._gcs_bucket is not None:
+      # Make sure the path ends with a /, otherwise GCS just returns the
+      # directory as a prefix and not list the contents.
+      prefix = self._path
+      if not self._path.endswith('/'):
+        prefix = f'{self._path}/'
+
+      # Unfortunately GCS doesn't work like a normal file system and the client
+      # library doesn't even pretend there is a directory hierarchy.
+      # The list API does return a list of prefixes that we can join with the
+      # list of objects to get something close to listdir(). But client library
+      # is pretty weird and it stores the prefixes on the iterator...
+      # https://github.com/googleapis/python-storage/blob/64edbd922a605247203790a90f9536d54e3a705a/google/cloud/storage/client.py#L1356
+      it = self._gcs_bucket.list_blobs(prefix=prefix, delimiter='/')
+      paths = [f.name for f in it] + [p.removesuffix('/') for p in it.prefixes]
+      r = [p.removeprefix(prefix) for p in paths]
+      return r
+
+    return os.listdir(self._path)
+
+  def exists(self) -> bool:
+    """exists returns true if the path is a file or directory."""
+    if self._gcs_bucket is not None:
+      return self.isfile() or self.isdir()
+
+    return os.path.exists(self._path)
+
+  def isfile(self) -> bool:
+    """isfile returns true if the path is a file."""
+    if self._gcs_bucket is not None:
+      return self._gcs_bucket.blob(self._path).exists()
+
+    return os.path.isfile(self._path)
+
+  def isdir(self) -> bool:
+    """isfile returns true if the path is a directory."""
+    if self._gcs_bucket is not None:
+      return len(self.listdir()) > 0
+
+    return os.path.isdir(self._path)
+
+  def makedirs(self):
+    """makedirs create parent(s) and directory in specified path."""
+    if self._gcs_bucket is not None:
+      # Do nothing. GCS doesn't have directories and files can be created with
+      # any path.
+      return
+
+    os.makedirs(self._path)
+
+  def open(self, *args, **kwargs) -> io.IOBase:
+    """
+    open returns a file handle to the file located at the specified path.
+
+    It has identical function signature to standard library open().
+    """
+    if self._gcs_bucket is not None:
+      return self._gcs_bucket.blob(self._path).open(*args, **kwargs)
+
+    return open(self._path, *args, **kwargs)
+
+  def getsize(self) -> int:
+    """getsize returns the byte size of the file at the specified path."""
+    if self._gcs_bucket is not None:
+      blob = self._gcs_bucket.get_blob(self._path)
+      if blob is None:
+        raise FileNotFoundError(
+            'GCS blob not found gs://{self._gcs_bucket.bucket}/{self._path}.')
+
+      # size can be None if use Bucket.blob() instead of Bucket.get_blob(). The
+      # type checker doesn't know this and insists we check if size is None.
+      return blob.size if blob.size is not None else 0
```

Comment:
nit: `blob.size if blob.size else 0` or even `blob.size or 0`

Reply:
Same as above. go/pystyle#truefalse-evaluations-decision it's better to have None comparisons explicit.